RandLAPACK/drivers/rl_cqrrp_gpu.hh

```diff
@@ -0,0 +1,804 @@
+        
+        if(this -> timing) {
+            cudaStreamSynchronize(strm);
+            nvtxRangePop();
+            copy_A_sk_t_stop = high_resolution_clock::now();
+            copy_A_sk_t_dur += duration_cast<microseconds>(copy_A_sk_t_stop - copy_A_sk_t_start).count();
+            nvtxRangePushA("piv_A_sk");
+            qrcp_piv_t_start = high_resolution_clock::now();
+        }
+
+        // Apply pivots to A_sk
+        RandLAPACK::cuda_kernels::col_swap_gpu(strm, sampling_dimension, cols, cols, A_sk_work, d, A_sk_copy_col_swap_work, d, J_buffer);
+
+        if(this -> timing) {
+            cudaStreamSynchronize(strm);
+            nvtxRangePop();
+            qrcp_piv_t_stop = high_resolution_clock::now();
+            qrcp_piv_t_dur += duration_cast<microseconds>(qrcp_piv_t_stop - qrcp_piv_t_start).count();
+        }
+
+        // Perform an unpivoted QR on A_sk
+        if(iter == 0) {
+            lapack::geqrf_work_size_bytes(sampling_dimension, cols, A_sk_work, d, &d_size_geqrf, &h_size_geqrf, lapack_queue);
+            d_work_geqrf = blas::device_malloc< char >( d_size_geqrf, lapack_queue );
+            std::vector<char> h_work_geqrf_vector( h_size_geqrf );
+            h_work_geqrf = h_work_geqrf_vector.data();
+        }
+        lapack::geqrf(sampling_dimension, cols, A_sk_work, d, Work2, d_work_geqrf, d_size_geqrf, h_work_geqrf, h_size_geqrf, d_info, lapack_queue);
+        
+        if(this -> timing) {
+            lapack_queue.sync();
+            nvtxRangePop();
+            qrcp_t_stop = high_resolution_clock::now();
+            qrcp_t_dur += duration_cast<microseconds>(qrcp_t_stop - qrcp_t_start).count();
+            nvtxRangePushA("copy_A");
+            copy_A_t_start = high_resolution_clock::now();
+        }
+        // THIS IS ALREADY DONE ABOVE
+        // Need to premute trailing columns of the full R-factor.
+        // Remember that the R-factor is stored the upper-triangular portion of A.
+        // Pivoting the trailing R and the ``current'' A.      
+        //RandLAPACK::cuda_kernels::copy_mat_gpu(strm, m, cols, &A[lda * curr_sz], lda, A_copy_col_swap, lda, false);    
+        
+        copy_queue.sync();
+        if(this -> timing) {
+            nvtxRangePop();
+            copy_A_t_stop = high_resolution_clock::now();
+            copy_A_t_dur += duration_cast<microseconds>(copy_A_t_stop - copy_A_t_start).count();
+            nvtxRangePushA("piv_A");
+            piv_A_t_start = high_resolution_clock::now();
+        }
+
+        RandLAPACK::cuda_kernels::col_swap_gpu(strm, m, cols, cols, &A[lda * curr_sz], lda, A_copy_col_swap, lda, J_buffer);
+        
+        // Checking for the zero matrix post-pivoting is the best idea, 
+        // as we would only need to check one column (pivoting moves the column with the largest norm upfront)
+        block_zero = true;
+        RandLAPACK::cuda_kernels::all_of(strm, rows, 0.0, A_work, block_zero);
```

Comment:
Please make sure this kernel is tolerance-based.

Reply:
DOne.